sdmetrics/single_table/equalized_odds.py

```diff
@@ -0,0 +1,465 @@
+# flake8: noqa
+"""EqualizedOddsImprovement metric for single table datasets."""
+
+import pandas as pd
+from sklearn.metrics import confusion_matrix
+
+from sdmetrics.goal import Goal
+from sdmetrics.single_table.base import SingleTableMetric
+from sdmetrics.single_table.utils import (
+    _validate_classifier,
+    _validate_column_consistency,
+    _validate_column_values_exist,
+    _validate_data_and_metadata,
+    _validate_prediction_column_name,
+    _validate_required_columns,
+    _validate_sensitive_column_name,
+    _validate_tables,
+    _process_data_with_metadata_ml_efficacy_metrics,
+)
+
+
+class EqualizedOddsImprovement(SingleTableMetric):
+    """EqualizedOddsImprovement metric.
+
+    This metric evaluates fairness by measuring equalized odds - whether the
+    True Positive Rate (TPR) and False Positive Rate (FPR) are the same
+    across different values of a sensitive attribute.
+
+    The metric compares the equalized odds between real training data and
+    synthetic data, both evaluated on a holdout validation set.
+    """
+
+    name = 'EqualizedOddsImprovement'
+    goal = Goal.MAXIMIZE
+    min_value = 0.0
+    max_value = 1.0
+
+    @classmethod
+    def _validate_data_sufficiency(
+        cls,
+        data,
+        prediction_column_name,
+        sensitive_column_name,
+        positive_class_label,
+        sensitive_column_value,
+    ):
+        """Validate that there is sufficient data for training."""
+        # Create binary versions of the columns
+        prediction_binary = data[prediction_column_name] == positive_class_label
+        sensitive_binary = data[sensitive_column_name] == sensitive_column_value
+
+        # Check both sensitive groups (target value and non-target value)
+        for is_sensitive_group in [True, False]:
+            group_predictions = prediction_binary[sensitive_binary == is_sensitive_group]
+            group_name = 'sensitive' if is_sensitive_group else 'non-sensitive'
+
+            if len(group_predictions) == 0:
+                raise ValueError(f'No data found for {group_name} group.')
+
+            positive_count = group_predictions.sum()
+            negative_count = len(group_predictions) - positive_count
+
+            if positive_count < 5 or negative_count < 5:
+                raise ValueError(
+                    f'Insufficient data for {group_name} group: {positive_count} positive, '
+                    f'{negative_count} negative examples (need ≥5 each).'
```

Comment:
Do we need a for loop here since we are counting both positive and negative?

Reply:
Yes. There are two groups (1) data that matches the sensitive value and (2) data that doesn't. For both cases we need at least 5 True target values and 5 False target values.